{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1>Binnacle-Estimating Scaffold Coverages</h1>\n",
    " \n",
    " This repository is under active development. In this document we describe the various functions implemented in **Compute\\_Scaffold\\_Coverages\\_Utility.py** and **Binnacle\\_IO\\_Utility.py**.\n",
    " \n",
    "In the following section, we describe all the functions associated with **Compute\\_Scaffold\\_Coverages\\_Utility.py**. \n",
    "\n",
    "<h2>Computing Global Coordinates</h2>\n",
    "\n",
    "**Function**: Compute\\_Global\\_Coordinates <br/>\n",
    "**Inputs**: subgraph of the weakly connected component, available vertex v<br/>\n",
    "**Outputs**: dictionary describing the global coordinates for each vertex<br/>\n",
    "\n",
    "<ol> \n",
    "    <li>This function computes the global coordinate system for the scaffold(connected component) based on the length of the contigs in the connected component and the linking information as returned by MetaCarvel. We assign (0, len(u)) where u is an \"available vertex\" of the subgraph. Available vertex is any vertex in the subgraph that has indegree 0. For all other vertices we assign coordinates in the global coorinate system in breadth first manner.</li>   \n",
    "    <li>We also consider the contig orientations and the edge orientations in accurate estimation of coordinates. </li>\n",
    "    <li>If there multiple possible assignments we pick the one that has the largest coordinate value. We choose the greedy approaach because the number of solutions grows exponentially and the problem is NP-Hard! </li>\n",
    "    <li>Finally, we normalize all other positions based on the least cooridnate value so that the coorinate system starts with 0.</li>\n",
    "    <li>The input to the program is the connected component subgraph of the assembly graph \"oriented.gml\" which is a result of running MetaCarvel.</li>\n",
    "</ol>\n",
    "\n",
    "\n",
    "\n",
    "<h2>Computing the Depth of Coverage along the Scaffold's Global Coordinate System </h2>\n",
    "\n",
    "**Function**: Compute_Coverage<br/>\n",
    "**Input**: subgraph of the weakly connected component, available vertex u and the dataframe containing the coverage of each contig <br/>\n",
    "**Output**: Coverage of the scaffold and per contigs depth<br/>\n",
    "\n",
    "To compute the depth we run the Compute_Global_Coordinates and Load_Read_Coverage functions and pass this to the Compute_Coverage described below. This program estimates the per cooridnate depth by using the outputs of the said two information. The program returns the coverage along the entire scaffold and the coverage along the longest path and a dataframe contiaining the contig, its average depth and its coordinates in the scaffold.\n",
    "\n",
    "<h2> Change Point Detection on the Scaffold Coverage </h2>\n",
    "\n",
    "**Function**: Helper_Changepoints<br/>\n",
    "**Input**: Coverage Vector, Window_Size(Default 1500bp)<br/>\n",
    "**Output**: Change Point Statistic Vector<br/>\n",
    "\n",
    "**Function**: Helper_Changepoints_Z_Stat<br/>\n",
    "**Input**: Coverage Vector, Window_Size(Default 1500bp)<br/>\n",
    "**Output**: Change Point Statistic Vector<br/>\n",
    "\n",
    "To compute changepoints along the scaffold we slide a window of default size 1500 along the scaffold and take the ratios of the maximum of means of the window preceeding any point and the window suceeding the point and the minimum of the said quantities. Any abnromal spike magnitude indicates a change point. If the size of the contig is less than the size of the window we adaptively pick a smaller window. \n",
    "\n",
    "<h2> Outlier Detection in Change Point Signals </h2>\n",
    "\n",
    "**Function**: ID_Outliers<br/>\n",
    "**Input**: Change Point Statistic Vector, threshold<br/>\n",
    "**Output**: Outlier Indices<br/>\n",
    "\n",
    "**Function**: ID_Peaks<br/>\n",
    "**Input**: Change Point Statistic Vector, threshold <br/>\n",
    "**Outputs**: Outlier Indices<br/>\n",
    "\n",
    "**Function**: Filter_Neighbors<br/>\n",
    "**Input**: Outlier List, Change Point Statistic Vector and Vicinity Window Size<br/>\n",
    "**Output**: Filtered Outlier List<br/>\n",
    "\n",
    "<ol>\n",
    "    <li>The following code segment is used to identify outliers in change points. To do so, we compute the peaks first.</li> \n",
    "    <li>We don't directly identify outliers on the signal because doing so would pickup points near the peaks, these are indicators of sliding windows and not really outliers. Our change point statistic is computed as follows. Let $\\mu_i$ and $\\mu_{i+1}$ the means of the subsequent windows, out statistic is given by $$\\rho = \\frac{max(\\mu_i, \\mu_{i+1})}{min(\\mu_i, \\mu_{i+1})}$$</li> \n",
    "    <li>To overcome the issue, we first identify peaks. A point is a peak if a point is larger than its predecessor and successor. This is performed by the function *ID_Peaks*.</li> \n",
    "    <li>The output of this passed to *ID_outliers* which picks all those points that is gretaer than the point which is the *thresh*'s percentile. The default value of *thresh* is 98.5. </li>\n",
    "    <li>The filter outliers is still a work in progress. This is aimed at removing all the outliers points that is close to one another, but this is not super important. While the mthod described in the following block is data driven we are working on improving this method by examining the underlying graph structure.</li>\n",
    "</ol>\n",
    "\n",
    "<h2> Rules for delinking contigs at change points. </h2> \n",
    "\n",
    "**Function**: Get_Outlier_Contigs<br/>\n",
    "**Input**: Outlier Positions, Subgraph of the Scaffold, Position cutoff to take action on the contig, and a dictionary describing the contig pileup at each position. <br/>\n",
    "**Output**: subgraph after stratification. <br/>\n",
    "\n",
    "The rules for delinking the contigs are described below. The first row represent the orientation of the contig and the first colum represent the location of the contig relative to the contig at the change point. \n",
    "\n",
    "|#                   |  Forward            |  Reverse          |\n",
    "|:------------------:|---------------------|-------------------|\n",
    "|        Start       | Delink predecessor  | Delink successor  |\n",
    "|          End       | Delink successor    | Delink predecessor|\n",
    "\n",
    "The input to the program is the set of outliers, the dictionary containing the contigs at each position, a vector of means for detecting change points and a *pos_cutoff* that specifies the number of basepairs from the contig ends to considered for delinking.\n",
    "\n",
    "<h2> Contig Pileup Dictionary </h2>\n",
    "\n",
    "**Function**: Return_Contig_Scaffold_Poistions<br/>\n",
    "**Input**: Global Coordinate Dictionary<br/>\n",
    "**Output**: Dictionary describing the pileup of contigs<br/>\n",
    "\n",
    "We iterate through the global coordinate dictionary and return a dictionary that has the postion along the scaffold as the key and pile up of contig along each position as its values.\n",
    "\n",
    "<h2> Computing an Available Vertex </h2>\n",
    "\n",
    "**Function**: Return_Starting_Point<br/>\n",
    "**Input**: Graph<br/>\n",
    "**Output**: vertex with the lowest indegree and the indegree value <br/>\n",
    "\n",
    "The above function returns the node with lowest indegree and the indegree value. If the minimum indegree is 0, then there is an available node and the program procceds by computing the global coordinates. If not we simplify the graph\n",
    "\n",
    "<h2> Handling Non-DAG Elements </h2>\n",
    "\n",
    "**Function**: Random_Simplify<br/>\n",
    "**Input**: Graph<br/>\n",
    "**Output**: Stratified Graph <br/>\n",
    "\n",
    "Most of the scaffolds in the sample form a Directed Acyclic Graph. Some of the induce a cycle. Even if the scaffold has a cycle we look for an available vertex if not, the above code fragment converts the graph into a DAG by randonly snipping an edge in the simple cycle(s) in the scaffold. \n",
    "\n",
    " \n",
    " \n",
    "In the following section we describe the code fragments available in **Binnacle_IO_Utility.py** which deals with loading and processing the function described in this section.\n",
    "\n",
    "<hr />\n",
    "\n",
    "<h2> Loading Assembly Graph and Coverages </h2>\n",
    "\n",
    "**Function**: Load_Read_Coverage<br/>\n",
    "**Input**: Path to the output of running genomecov, List of nodes in the assembly graph, Output directory<br/>\n",
    "**Output**: Return a networkx graph object and dataframe containing the output of genomecov\n",
    "\n",
    "To compute the read coverages we use the *genomecov* program, part of the *bedtools* suite. We run *genomecov* with *-d* optional enabled so that we get the per base depth. The output of the prgram is utilized to compute the depth along the scaffold and this function loads the out of genomecov as a dataframe.   \n",
    "\n",
    "\n",
    "<h2> Computing Coverages and Writing the Results </h2> \n",
    "\n",
    "**Function**: Write_Coverage_Outputs<br/>\n",
    "**Input** : networkx object of assembly graph, dataframe of coverage, output directory to dump outputs <br\\>\n",
    "**Output**: A tab seperated text file for each of the following, \n",
    "<ol>\n",
    "    <li>Global Coordinates for the Scaffold Before Delinking with the Headers: \n",
    "        <ol><li>Scaffold id</li> \n",
    "            <li>Contig id</li> \n",
    "            <li>Start position in the global coordinate system</li> \n",
    "            <li>End position of the contig in the global coordinate system </li></ol>\n",
    "    </li>\n",
    "    <li>Global Coordinates for the Scaffold After Delinking with the Headers: \n",
    "        <ol><li>Scaffold id after delinking</li> \n",
    "            <li>Scaffold id before delinking</li> \n",
    "            <li>Contig id</li> \n",
    "            <li>Start position in the global coordinate system</li> \n",
    "            <li>End position of the contig in the global coordinate system</li></ol>\n",
    "    </li>\n",
    "    <li>Scaffold Coverages Before Delinking with the Headers: \n",
    "        <ol><li>Scaffold id before delinking</li> \n",
    "            <li>Location in the scaffold</li> \n",
    "            <li>coverage Value</li></ol>\n",
    "    </li>\n",
    "    <li>Scaffold Coverages After Delinking with the Headers: \n",
    "        <ol><li>Scaffold id after delinking</li> \n",
    "            <li>Scaffold id before delinking</li> \n",
    "            <li>Location in the scaffold</li> \n",
    "            <li>coverage Value</li></ol>\n",
    "    </li>\n",
    "    <li>Scaffold Level cover Summary Before Delinking with the Headers: \n",
    "        <ol><li>Scaffold id before delinking</li> \n",
    "            <li>Mean Coverage</li> \n",
    "            <li>Deviation in Coverage</li></ol>\n",
    "    </li>\n",
    "    <li>Scaffold Level cover Summary After Delinking with the Headers: \n",
    "        <ol><li>Scaffold id After delinking</li> \n",
    "            <li>Mean Coverage</li> \n",
    "            <li>Deviation in Coverage</li></ol>\n",
    "    </li>\n",
    "    \n",
    "</ol>\n",
    "\n",
    "The function imports the **Compute\\_Scaffold\\_Coverages\\_Utility.py** and uses the function create the above output files. \n",
    "    \n",
    "\n",
    "<h2> Creating Scaffold.fasta Files </h2>\n",
    "\n",
    "**Function**: Load_FASTA_File<br/>\n",
    "**Input**: Path to the fasta file containing the contig sequences. <br/>\n",
    "**Output**: A dictionary containing the contig id as the keys and the sequences as values <br/>\n",
    "\n",
    "**Function**: Get_Contigs_in_Scaffolds<br/>\n",
    "**Input**: Path to the tsv file containing global coordinate information after delinking which is an output of running the Write_Coverage_Outputs<br/>\n",
    "**Output**: A dictionary containing the scaffold id as the keys and the contigs in the scaffold as values <br/>\n",
    "\n",
    "**Function**:  Write_Scaffolds<br/>\n",
    "**Input**: Path to the fasta file of the contigs, Path to the tsv file containing global coordinate information after delinking which is an output of running the Write_Coverage_Outputs, output directory to write the scaffolds <br/>\n",
    "**Output**: Writes the scaffolds to \"Scaffolds.fasta\" in the specified directory <br/>\n",
    "\n",
    "The function  Write_Scaffolds calls the above tw methods to create the scaffold.fasta file. \n",
    " * * *\n",
    "\n",
    "<h2> Computing Feature Matrices </h2> \n",
    "\n",
    "These functions compute all vs all alignment matrices and feature matrices. \n",
    "\n",
    "**Function**: Process_Scaffold_Coverages <br/>\n",
    "**Input**: Dataframe containing the coords obtained by running binnacle and the coverage dataframe obtained by load genomecov. This dataframe could any coverage output generated for peforming all vs all alignments. <br/>\n",
    "**Output**: Returns the summary of the coverage for each scaffold in the assembly graph. <br/>\n",
    "\n",
    "**Function**:Prepare_Feature_Matrix<br/>\n",
    "**Input**: Path to the coordinate file, the summary generated by binnacle on the parent coverage outputs, flag that indicates to perform all vs all alignments, directory that contains all the all vs all alignment for the sample, Location to write the output to <br/>\n",
    " * * *\n",
    "<h2> Visualize the Results </h2>\n",
    "This is under development....\n",
    "\n",
    "For questions contact hsmurali@umd.edu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
